src/conv-core/isomalloc.C

```diff
@@ -83,2113 +73,2072 @@ static int _mmap_probe = 0;
+    {
+      IMP_DBG("[%d][%p] RegionHeader::setOccupied(%u)\n", CmiMyPe(), this, (unsigned int)o);
 
-static void free_slots(slotset *ss, CmiInt8 sslot, CmiInt8 nslots)
-{
-  slotblock *sb_low  = find_btree_slotblock(ss->btree_root, sslot - 1);
-  slotblock *sb_high = find_btree_slotblock(ss->btree_root, sslot + nslots);
+      occupiedfield = o;
+    }
 
-  if (sb_low == NULL)
-  {
-    if (sb_high == NULL)
+    bool isFirst() const
     {
-      /* there is no adjacent slotblock, so create a new one and */
-      /* insert it in the b-tree */
-      ss->btree_root = btree_insert(ss, ss->btree_root, sslot, nslots);
+      return prev() == nullptr;
     }
-    else
+    bool isLast() const
     {
-      /* there is an adjacent slotblock to the right, so update its range */
-      CmiInt8 old_nslots = sb_high->nslots;
-      sb_high->startslot = sslot;
-      sb_high->nslots   += nslots;
-      list_move(ss, sb_high->listblock, old_nslots);
+      return next() == nullptr;
     }
-  }
-  else
-  {
-    if (sb_high == NULL)
+
+    bool prevIsEmpty() const
     {
-      /* there is an adjacent slotblock to the left, so update its range */
-      CmiInt8 old_nslots  = sb_low->nslots;
-      sb_low->nslots     += nslots;
-      list_move(ss, sb_low->listblock, old_nslots);
+      return !isFirst() && prev()->isEmpty();
     }
-    else
+    bool nextIsEmpty() const
     {
-      /* there are adjacent slotblocks on both sides (i.e., the
-         slots to be freed exactly span the gap between 2 slotblocks),
-         so update the range of the lower slotblock and delete the
-         upper one */
-      CmiInt8 old_nslots = sb_low->nslots;
-      sb_low->nslots     = sb_low->nslots + nslots + sb_high->nslots;
-      list_move(ss, sb_low->listblock, old_nslots);
-      ss->btree_root = btree_delete(ss, ss->btree_root, sb_high->startslot);
+      return !isLast() && next()->isEmpty();
     }
-  }
-}
-
-/*****************************************************************
- * Recursively free the allocated memory of the b-tree
- *****************************************************************/
 
-static void delete_btree(btreenode *node)
-{
-  for (int i = 0; i <= node->num_blocks; i++)
-  {
-    if (node->child[i] != NULL)
+    size_t size() const
     {
-      delete_btree(node->child[i]);
-      free_reentrant(node->child[i]);
+      return sizefield;
     }
-    else
+    size_t usersize() const
     {
-      return;
+      return size() - sizeof(RegionHeader);
     }
-  }
-}
 
-/*****************************************************************
- * Free the allocated memory of the list array
- *****************************************************************/
+  private:
+    RegionHeader * prevfield;
+    RegionHeader * nextfield;
+    size_t sizefield;
+    BitCarrier occupiedfield;
+  };
 
-static void delete_list_array(slotset *ss)
-{
-  for (int i = 0; i < LIST_ARRAY_SIZE; i++)
+  using Occupied = RegionHeader::Occupied;
+
+  struct RBTree
```

Comment:
std::map is almost always implemented as an rbtree. Could this be a simpler alternative?

Reply:
Using `std::map` or other STL containers would introduce calls to system malloc/free, which would have a performance impact. Is it semantically appropriate for an allocator to call another allocator on the critical path? Implementing our own structure has allowed me to tweak it to the needs of the mempool. For example, moving the RBTree code into the class definition so it could be inlined by the compiler gave a 9% speedup. Additionally, migration does not need to reconstruct the tree structure, since the tree is embedded in the mmapped region.

Reply:
Ok. It might be useful to add this as a comment somewhere.

Reply:
Added as a comment